src/codegen/codegen_compatibility_visitor.cpp

```diff
@@ -0,0 +1,142 @@
+                   it->is_independent_block()) {
+            std::string node_type_name = it->get_node_type_name();
+            // remove "Block" substring
+            node_type_name.erase(node_type_name.end() - 5, node_type_name.end());
+            // transform name of node type to capital letters, as it is in mod files
+            std::transform(node_type_name.begin(),
+                           node_type_name.end(),
+                           node_type_name.begin(),
+                           ::toupper);
+
+            ss << node_type_name << " construct found at [";
+            ss << it->get_token()->position() << "] is not supported\n";
+        } else if (it->is_function_table_block()) {
+            ss << "\"" << it->get_node_name() << "\" FUNCTION_TABLE construct found at [";
+            ss << it->get_token()->position() << "] is not supported\n";
+        } else if (it->is_global_var()) {
+            auto global_var = dynamic_cast<ast::GlobalVar*>(it.get());
+            if (node->get_symbol_table()->lookup(global_var->get_node_name())->get_write_count() >
+                0) {
+                ss << "\"" << global_var->get_node_name() << "\" variable found at [";
+                ss << global_var->get_token()->position();
+                ss << "] should be defined as a RANGE variable instead of GLOBAL to enable ";
+                ss << "backend transformations\n";
+            }
+        } else if (it->is_pointer_var()) {
+            ss << "\"" << it->get_node_name() << "\" POINTER found at [";
+            ss << it->get_token()->position();
+            ss << "] should be defined as BBCOREPOINTER to use it in CoreNeuron\n";
+        } else if (it->is_bbcore_pointer_var()) {
+            auto verbatim_nodes = AstLookupVisitor().lookup(node, AstNodeType::VERBATIM);
+            auto found_bbcore_read = false;
+            auto found_bbcore_write = false;
+            for (auto it: verbatim_nodes) {
+                auto verbatim = dynamic_cast<ast::Verbatim*>(it.get());
+                auto verbatim_statement = verbatim->get_statement();
+                auto verbatim_statement_string = verbatim_statement->get_value();
+                auto bbcore_read_string = verbatim_statement_string.find("bbcore_read");
+                auto bbcore_write_string = verbatim_statement_string.find("bbcore_write");
+                if (bbcore_read_string != std::string::npos) {
+                    found_bbcore_read = true;
+                }
+                if (bbcore_write_string == std::string::npos) {
+                    found_bbcore_write = true;
+                }
+            }
+            if (!found_bbcore_read) {
+                ss << "\"bbcore_read\" function not defined in any VERBATIM block\n";
+            }
+            if (!found_bbcore_write) {
+                ss << "\"bbcore_write\" function not defined in any VERBATIM block\n";
+            }
+        }
+    }
```

Comment:
Looking at how we construct error messages, I think it would be cleaner to define error codes now. I was thinking about this for quite some time but didnt get chance to implement it.

Overall idea would be : we have defined error codes with their messages. In the visitor we just collect error code and associated node. And then some `print_error` function take care of printing them in nice format.

In this function, the stream manipulation and other logic gets mixed up.

Reply:
I agree, but maybe we can defer this for later?